src/relax/op/op.cc

```diff
@@ -315,6 +315,32 @@ Expr MakeShapeOf(Expr expr) {
 
 TVM_REGISTER_GLOBAL("relax.op.shape_of").set_body_typed(MakeShapeOf);
 
+// tensor_to_shape
+
+StructInfo ReturnTensorToShapeStructInfo(const Call& call, const BlockBuilder& ctx) {
+  ICHECK(call->args.size() == 1);
+  ICHECK(call->args[0]->struct_info_.defined());
+  const auto* tsinfo = GetStructInfoAs<TensorStructInfoNode>(call->args[0]);
+  ICHECK(tsinfo && tsinfo->shape.defined());
```

Comment:
I think this condition is more restrictive than necessary. I think it's okay for the shape not to be known at compile time and to return `ShapeStructInfo(ndim=-1)` if the input rank/shape is unknown at compile time.

Reply:
Do you mean these lines?
```
  // define symbolic variables
  Array<PrimExpr> shape_var;
  for (int i = 0; i < sinfo->ndim; i++) {
    shape_var.push_back(tir::Var("x", DataType::Int(64)));
  }
```
Initially, I also wanted to support the unknown rank but realized it is trickier than I thought. 
The problem is you need to insert these symbolic variables at compile-time so we need this info.

Reply:
What exactly are the new shape vars needed for?

Reply:
To MatchCast and put them in the `ShapeExpr`.
```
// define symbolic variables
Array<PrimExpr> shape_var;
for (int i = 0; i < sinfo->ndim; i++) {
    shape_var.push_back(tir::Var("x", DataType::Int(64)));
 }

// bind symbolic variables to the shape tuple
relax::Var var("y", ShapeStructInfo(shape_var));
builder_->EmitNormalized(MatchCast(var, call, ShapeStructInfo(shape_var)));
return ShapeExpr(shape_var);
```

Reply:
That much I understand, what I am more curious about is why we need to construct this shape expression

Reply:
So that later reshape op can take it as the second input. 
```
def reshape(x: Expr, shape: Union[Tuple[PrimExprLike], Expr]) -> Expr
```
Is there any better way to do this?

Reply:
Can't we just bind the result of `tensor_to_shape` to a variable and pass that variable as the shape argument to `reshape`?

Reply:
The contract with lowering is that we need to know each shape value (could be symbolic var) for memory pre-allocation. 
So, this example wouldn't lower. 
```
@I.ir_module
class Module:
    @R.function
    def main(x: R.Tensor((3,), dtype="int64")) -> R.Tensor((3,), dtype="int64"):
        gv: R.Shape(ndim=3) = R.call_packed(
            "vm.builtin.tensor_to_shape", x, sinfo_args=(R.Shape(ndim=3),)
        )
        gv3: R.Tensor(ndim=3, dtype="int64") = R.reshape(x, gv)
        gv_1: R.Tensor(ndim=3, dtype="int64") = gv3
        return gv_1
mod = LegalizeOps()(Module)
mod.show()
```

Reply:
great discussion folks! Looks the limitation for shape_expr is from TE, interesting

Reply:
Hmmmm, I think we should put this on the list to fix later. Let's please note this limitation in a comment, since it is not obvious from looking at the code

Reply:
Added this in the comment. Would it be good to go now?